compiler/noirc_evaluator/src/ssa/opt/expand_signed_checks.rs

```diff
@@ -0,0 +1,606 @@
+use acvm::{FieldElement, acir::AcirField};
+
+use crate::ssa::{
+    ir::{
+        function::Function,
+        instruction::{Binary, BinaryOp, ConstrainError, Instruction},
+        types::{NumericType, Type},
+        value::ValueId,
+    },
+    ssa_gen::Ssa,
+};
+
+use super::simple_optimization::SimpleOptimizationContext;
+
+impl Ssa {
+    /// Expands signed arithmetic operations to include explicit overflow checks.
+    ///
+    /// See [`expand_signed_checks`][self] module for more information.
+    #[tracing::instrument(level = "trace", skip(self))]
+    pub(crate) fn expand_signed_checks(mut self) -> Ssa {
+        for function in self.functions.values_mut() {
+            function.expand_signed_checks();
+        }
+        self
+    }
+}
+
+impl Function {
+    /// The structure of this pass is simple:
+    /// Go through each block and re-insert all instructions, decomposing any checked signed arithmetic to have explicit
+    /// overflow checks.
+    pub(crate) fn expand_signed_checks(&mut self) {
+        // TODO: consider whether we can implement this more efficiently in brillig.
+
+        self.simple_reachable_blocks_optimization(|context| {
+            let instruction_id = context.instruction_id;
+            let instruction = context.instruction();
+
+            // We only care about binary instructions that are signed arithmetic operations...
+            let Instruction::Binary(Binary {
+                lhs,
+                rhs,
+                operator:
+                    operator @ (BinaryOp::Add { unchecked: false }
+                    | BinaryOp::Sub { unchecked: false }
+                    | BinaryOp::Mul { unchecked: false }),
+            }) = instruction
+            else {
+                return;
+            };
+
+            // ... and it must be a signed integer operation.
+            if !context.dfg.type_of_value(*lhs).is_signed() {
+                return;
+            }
+
+            let lhs = *lhs;
+            let rhs = *rhs;
+            let operator = *operator;
+
+            // We remove the current instruction, as we will need to replace it with multiple new instructions.
+            context.remove_current_instruction();
+
+            let old_result = *context.dfg.instruction_results(instruction_id).first().unwrap();
+
+            let mut simple_context = Context { context };
+            let new_result = match operator {
+                BinaryOp::Add { .. } => simple_context.insert_add(lhs, rhs),
+                BinaryOp::Sub { .. } => simple_context.insert_sub(lhs, rhs),
+                BinaryOp::Mul { .. } => simple_context.insert_mul(lhs, rhs),
```

Comment:
```suggestion
            let mut expansion_context = Context { context };
            let new_result = match operator {
                BinaryOp::Add { .. } => expansion_context.insert_add(lhs, rhs),
                BinaryOp::Sub { .. } => expansion_context.insert_sub(lhs, rhs),
                BinaryOp::Mul { .. } => expansion_context.insert_mul(lhs, rhs),
```
I see `context` is a `SimpleOptimizationContext` type, which is where I think the name `simple_context` comes from the wrapper, but I'm not sure why this `Context` type is considered _simple_.